### PR TITLE
WITH clause for Insert, Update, Delete

### PR DESCRIPTION
**Pull request by naithar**
Added a `with` methods for other queries other than `Select`.
Also added an `.on` for `.join` in `Select`'s `with` test to workaround PostgreSQL not being able to perform the result query.
Also moved `with` tests to corresponding query test file.

Issue: #40

### CODE REVIEW

File: Sources/SwiftKuery/Delete.swift [hunk 1]
```diff
@@ -28,6 +28,9 @@ public struct Delete: Query {
     /// A String with a clause to be appended to the end of the query.
     public private (set) var suffix: QuerySuffixProtocol?
     
+    /// An array of `AuxiliaryTable` which will be used in WITH clause.
```
Comment (irar2) [inline Sources/SwiftKuery/Delete.swift]: Please change to: An array of `AuxiliaryTable` which will be used in **a query with a** WITH clause.

Please do this in Insert, Select, and Update as well

**Comment by naithar** [comment]
@irar2 made changes that's requested.

**Comment by irar2** [comment]
Thanks

**Comment by irar2** [comment]
@naithar This doesn't work on Postgres.
I only tried Update, and it requires `from withTable` to be added, i.e. a query (similar to the test in TestUpdate):
```sql
WITH tmp AS (SELECT a AS d FROM t1) 
UPDATE t2 SET t2.c='kiwi' 
WHERE tmp.d!=t2.c;
```
gets Postgres error:
ERROR:  missing FROM-clause entry for table "tmp"
LINE 1: ...t a as d from t1) update t2 set t2.c='kiwi' where tmp.d!=t2....

It should be 
```sql
WITH tmp AS (SELECT a AS d FROM t1) 
UPDATE t2 SET t2.c='kiwi' 
FROM tmp
WHERE tmp.d!=t2.c;
```
Swift-Kuery.Update doesn't have `from`, so with this PR we generate invalid queries for Postgres.

**Comment by naithar** [comment]
@irar2 sqlite doesn't have a `FROM` clause for `UPDATE`, so i guess the best way to fix this atm is to revert parts of the PR and implement an Extension so drivers could implement specific functionality,  like mentioned in #25

**Comment by naithar** [comment]
@irar2 [this commit](https://github.com/naithar/Swift-Kuery/commit/7c8cc3787aca770bce95d7ae376bf20c84070a6e) contains initial implementation of `Extensions` and `WITH` clause implementation for `Select` using extensions. 
Would it better to move `WITH` clauses to `Swift-Kuery-PostgreSQL` after and if this implementation of extensions is accepted?

**Comment by irar2** [comment]
@naithar Thanks, we took some time to evaluate your suggestion. We think, we understand your goal to generalize db specific extensions. This is obviously a good idea. However, we are not sure your injection points are sufficient, and also the use of priorities in case of several extensions at the same position worries us. Maybe we can start an issue to discuss how we should handle db differences in general.

It's probably better to keep WITH in Swift-Kuery, since we would like to enable Swift-Kuery to handle as much as possible and enable the user to switch between dbs. 

In the particular case of WITH and the need in FROM clause in Postgres, maybe we can always add FROM clause automatically for WITH (this will require changes in Update and Delete, of course), and add a flag in QueryBuilder to determine whether to include it or not. This can be a temporary solution until we figure out what to do in general. What do you think?

**Comment by naithar** [comment]
@irar2 sure, an issue to discuss an extension feature would be a good starting point. An implementation that i referenced was just my view on the problem.

Keeping `WITH` (or any other feature) in `Swift-Kuery` could bring a problem if it's used in db that does not support it. For example `WITH` is not supported by MySQL as far as i know.

But as a temporary workaround, since MySQL driver is not implemented, flagging `QueryBuilder` can work. But a way to flag it should also be discussed.

**Comment by irar2** [comment]
[MySQL 8.0.1](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-1.html) will support WITH: 
"MySQL now supports common table expressions, both nonrecursive and recursive. Common table expressions enable use of named statement-local temporary tables, implemented by permitting a WITH clause preceding SELECT statements, and certain other statements."

**Comment by naithar** [comment]
@irar2 then I'll look into implementing a temporary workaround for Postrgres for `WITH` clause in Update-Delete.

**Pull request merged**